src/Microsoft.AspNetCore.OData/Formatter/ODataOutputFormatter.cs

```diff
@@ -223,11 +224,25 @@ public override Task WriteResponseBodyAsync(OutputFormatterWriteContext context,
                     };
                 };
 
+                Object value = context.Object;
+
+                // We create an ODataError where we don't have a SerializableError yet we have Non-Success Status Code.
+                if (!response.IsSuccessStatusCode() && type != typeof(SerializableError))
+                {
+                    type = typeof(ODataError);
+                    ODataError error = new ODataError()
+                    {
+                        Message = value.ToString(),
+                        ErrorCode = response.StatusCode.ToString()
+                    };
+                    value = error;
+                }
+
```

Comment:
I can't understand why you put codes here. We shall put the codes before going to output formatter.


Error handler is a project that we want to use a separate nuget package handler.

Reply:
@xuzhg 
We have inconsistencies in the way we handle non-success responses from an `ODataController`, e.g `NotFound("message")` or `BadRequest("message")`. For responses to be serialized as `ODataError`, the object value passed to the `WriteToStream` method, should be either of type `SerializableError` or `HttpError` or `ODataError`. See https://github.com/OData/WebApi/blob/800166239db589dfb606bb120f793f4aa75eef31/src/Microsoft.AspNet.OData.Shared/Formatter/ODataOutputFormatterHelper.cs#L107

In `AspNet` classic, we don't have a custom formatter `ODataOutputFormatter`, and the object value passed to the `WriteToStream` method, is of type `HttpError`. Below will be the response _(Not the missing error code)_

``` json
{
    "error": {
        "code": "",
        "message": "Error Message"
    }
}
```

In `AspNetCore`, the custom formatter `ODataOutputFormatter` doesn't handle some responses very well in the overriden method `WriteResponseBodyAsync`. https://github.com/OData/WebApi/blob/800166239db589dfb606bb120f793f4aa75eef31/src/Microsoft.AspNetCore.OData/Formatter/ODataOutputFormatter.cs#L189
When you return `NotFound("message")` or `BadRequest("message")` from the controller, the object value passed to the `WriteToStream` message, is of type `string`, not the expected `SerializableError` or `HttpError` or `ODataError`. Below is the response returned:

``` json
{
  "@odata.context":"https://localhost:XXXX/odata/$metadata#Edm.String",
  "value":"Error message"
}
```
In this fix, I decided to check the `StatusCode` in the `HttpResponse`. If it's non-success, I create an `ODataError` which returns the correct response as per the OData spec:

``` json
{
    "error": {
        "code": "400",
        "message": "Error Message"
    }
}
```

I welcome suggestions on how we can fix this in an optimal way.

Reply:
@xuzhg 
Updated the code as we discussed.
#Resolved